inputremapper/injection/consumer_control.py

```diff
@@ -93,23 +104,40 @@ async def run(self):
                 # won't appear, no need to forward or map them.
                 continue
 
+            for listener in self.context.listeners:
+                asyncio.create_task(listener(event))
+                # allow the just created task to do some stuff
+                await asyncio.sleep(0)
```

Comment:
According to the docstring in the old `delayed_handle_keycode` function this was, so that if_single gets a chance to first start its then/else child-macro before the key is used to do the next macro.

If `if_single` injects a modifier, this modifier should be active before the next handler injects an "a" or something, so that it is possible to capitalize it via `if_single`.

1. event from keyboard arrives
2. the asyncio.event if_single is waiting for triggers
3. if_single decides runs then
4. the original event is forwarded (or whatever it is supposed to do)

Reply:
The listeners are called before he handlers inject/forward anything, so your changes are probably correct

```suggestion
            for listener in self.context.listeners:
                asyncio.create_task(listener(event))
                
            # Running macros have priority, give them a head-start for processing the
            # event.  If if_single injects a modifier, this modifier should be active
            # before the next handler injects an "a" or something, so that it is
            # possible to capitalize it via if_single.
            # 1. Event from keyboard arrives (e.g. an "a")
            # 2. The asyncio.event if_single is waiting for triggers
            # 3. if_single decides runs then (e.g. injects shift_L)
            # 4. The original event is forwarded (or whatever it is supposed to do)
            # 5. Capitalized "A" is injected.
            # So make sure to call the listeners before notifying the handlers.
            await asyncio.sleep(0)
```

Reply:
good idea. But I still want it to be inside the for loop so that we give more time if there are more listeners, and don't sleep if there is nor reason to sleep